File: flask_app/app/services/membership_tier_service.py
from datetime import datetime
from app.dao.membership_tier_dao import MembershipTierDAO
from app.services.user_tier_metrics_service import UserTierMetricsService
from app.services.order_service import OrderService

class MembershipTierService:

    @staticmethod
    def get_all(page=1, per_page=10, filters=None):
        return MembershipTierDAO.get_all(page=page, per_page=per_page, filters=filters or {})

    @staticmethod
    def get_by_id(tier_id):
        return MembershipTierDAO.get_by_id(tier_id)

    @staticmethod
    def create(data):
        return MembershipTierDAO.create(data)

    @staticmethod
    def update(tier_id, data):
        return MembershipTierDAO.update(tier_id, data)

    @staticmethod
    def soft_delete(tier_id):
        return MembershipTierDAO.soft_delete(tier_id)
    
    @staticmethod
    def get_eligible_tiers(user_id):

        current_month_year = datetime.now().strftime("%Y-%m")
        monthly_metrics = UserTierMetricsService.get_for_user_month(
            user_id=user_id,
            month_year=current_month_year
        )

        min_monthly_orders = monthly_metrics.order_count if monthly_metrics else 0
        min_monthly_spend = monthly_metrics.total_spent if monthly_metrics else 0
        
        # users orders, spend from orders
        total_orders = OrderService.get_total_orders(user_id)
        total_spend = OrderService.get_total_spend(user_id)

        data = {
           "monthly_orders": min_monthly_orders,
            "monthly_spend": min_monthly_spend,
            "total_orders": total_orders,
            "total_spend": total_spend,
        }
        
        return MembershipTierDAO.get_eligible_tier(data)
    

File: springboot_api/src/main/java/com/example/membership_api/service/MembershipTierService.java
package com.example.membership_api.service;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.membership_api.dao.MembershipTierDAO;
import com.example.membership_api.dao.OrderDAO;
import com.example.membership_api.exception.NotFoundException;
import com.example.membership_api.model.MembershipTier;
import com.example.membership_api.model.UserTierMetrics;


@Service
public class MembershipTierService {

    @Autowired
    private MembershipTierDAO dao;

    @Autowired
    private UserTierMetricsService userTierMetricsService;

    @Autowired
    private OrderDAO orderDAO;

    public Map<String, Object> getAll(int page, int perPage, Map<String, Object> filters) {
        List<MembershipTier> items = dao.getAll(page, perPage, filters);
        long total = dao.getTotal(filters);
        return Map.of("items", items, "total", total, "page", page, "perPage", perPage);
    }

    public MembershipTier getById(String id) {
        MembershipTier tier = dao.getById(id);
        if (tier == null) throw new NotFoundException("Tier not found");
        return tier;
    }

    public List<MembershipTier> getActiveByTier(int tier){
        return dao.getActiveByTier(tier);
    }

    public MembershipTier create(Map<String, Object> tier) {
        return dao.create(tier);
    }

    public MembershipTier update(String id, Map<String, Object> updates) {
        MembershipTier existing = getById(id);
        return dao.update(existing, updates);
    }

    public boolean softDelete(String id) {
        return dao.softDelete(id);
    }

    public MembershipTier getEligibleTier(String userId) {
        String currentMonthYear = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM"));

        UserTierMetrics monthlyMetrics = userTierMetricsService.getForUserMonth(userId, currentMonthYear);

        int monthlyOrders = monthlyMetrics != null ? monthlyMetrics.getOrderCount() : 0;
        double monthlySpend = monthlyMetrics != null ? monthlyMetrics.getTotalSpent() : 0.0;

        long totalOrders = orderDAO.getTotalOrders(userId);
        double totalSpend = orderDAO.getTotalSpend(userId);

        Map<String, Object> data = new HashMap<>();
        data.put("monthly_orders", monthlyOrders);
        data.put("monthly_spend", monthlySpend);
        data.put("total_orders", totalOrders);
        data.put("total_spend", totalSpend);
        System.out.printf("getEligibleTier: %s", data.toString());
        return dao.getEligibleTier(data);
    }
}
